lib/kfifo-test.c

```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * KUnit test for the generic kernel FIFO implementation.
+ *
+ * Copyright (C) 2024 Diego Vieira <diego.daniel.professional@gmail.com>
+ */
+#include <kunit/test.h>
+
+#include <linux/kfifo.h>
+
+static void kfifo_test_reset_should_clear_the_fifo(struct kunit *test)
+{
+	DEFINE_KFIFO(my_fifo, u8, 32);
+
+	kfifo_put(&my_fifo, 1);
+	kfifo_put(&my_fifo, 2);
+	kfifo_put(&my_fifo, 3);
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 3);
+
+	kfifo_reset(&my_fifo);
+
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 0);
+	KUNIT_EXPECT_TRUE(test, kfifo_is_empty(&my_fifo));
+}
+
+static void kfifo_test_define_should_define_an_empty_fifo(struct kunit *test)
+{
+	DEFINE_KFIFO(my_fifo, u8, 32);
+
+	KUNIT_EXPECT_TRUE(test, kfifo_initialized(&my_fifo));
+	KUNIT_EXPECT_TRUE(test, kfifo_is_empty(&my_fifo));
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 0);
+}
+
+static void kfifo_test_len_should_ret_n_of_stored_elements(struct kunit *test)
+{
+	DEFINE_KFIFO(my_fifo, u8, 32);
+
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 0);
+
+	u8 *buffer1 = kunit_kmalloc_array(test, 5, sizeof(u8), GFP_KERNEL);
+	// Fill buffer1 with values 1, 2, 3, 4, 5
+	for (int i = 0; i < 5; i++) {
+		buffer1[i] = i + 1;
+	}
+
+	kfifo_in(&my_fifo, buffer1, 5);
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 5);
+
+	kfifo_in(&my_fifo, buffer1, 5);
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 10);
+
+	kfifo_reset(&my_fifo);
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 0);
+}
+
+static void kfifo_test_put_should_insert_and_get_should_pop(struct kunit *test)
```

Comment:
this function would be better with a for loop. you can statically define an array and then compare with the fifo output

```c
int arr[] = {3, 5, 11};

for (...) {
   if (out_data  == arr[i])
   ...
}
```